dbms/src/Flash/Mpp/ReceivedMessageQueue.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Mpp/ReceivedMessageQueue.h>
+
+namespace DB
+{
+namespace FailPoints
+{
+extern const char random_receiver_local_msg_push_failure_failpoint[];
+extern const char random_receiver_sync_msg_push_failure_failpoint[];
+extern const char random_receiver_async_msg_push_failure_failpoint[];
+} // namespace FailPoints
+
+namespace
+{
+void injectFailPointReceiverPushFail(bool & push_succeed [[maybe_unused]], ReceiverMode mode)
+{
+    switch (mode)
+    {
+    case ReceiverMode::Local:
+        fiu_do_on(FailPoints::random_receiver_local_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    case ReceiverMode::Sync:
+        fiu_do_on(FailPoints::random_receiver_sync_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    case ReceiverMode::Async:
+        fiu_do_on(FailPoints::random_receiver_async_msg_push_failure_failpoint, push_succeed = false);
+        break;
+    default:
+        RUNTIME_ASSERT(false, "Illegal ReceiverMode");
+    }
+}
+} // namespace
+
+template <bool fine_grained_shuffle, bool need_wait>
+std::pair<MPMCQueueResult, ReceivedMessagePtr> ReceivedMessageQueue::pop(size_t stream_id)
+{
+    MPMCQueueResult res;
+    ReceivedMessagePtr recv_msg;
+    if constexpr (fine_grained_shuffle)
+    {
+        if constexpr (need_wait)
+        {
+            res = msg_channels_for_fine_grained_shuffle[stream_id]->pop(recv_msg);
+        }
+        else
+        {
+            res = msg_channels_for_fine_grained_shuffle[stream_id]->tryPop(recv_msg);
+        }
+        if (recv_msg != nullptr)
+        {
+            if (recv_msg->remaining_consumers->fetch_sub(1) == 1)
+            {
+                ReceivedMessagePtr original_msg;
+                auto pop_result [[maybe_unused]] = grpc_recv_queue->tryPop(original_msg);
+                /// if there is no remaining consumer, then pop it from original queue, the message must stay in the queue before the pop
+                /// so even use tryPop, the result must not be empty
+                assert(pop_result != MPMCQueueResult::EMPTY);
+                if (original_msg != nullptr)
+                    RUNTIME_CHECK_MSG(*original_msg->remaining_consumers == 0, "Fine grained receiver pop a message that is not full consumed, remaining consumer: {}", *original_msg->remaining_consumers);
```

Comment:
```suggestion
                if (original_msg != nullptr)
                    RUNTIME_CHECK_MSG(original_msg != nullptr && *original_msg->remaining_consumers == 0, "Fine grained receiver pop a message that is not full consumed, remaining consumer: {}", *original_msg->remaining_consumers);
```

Reply:
The result could be `Cancelled`, so `original_msg ` is not always not null

Reply:
my mistake...
```suggestion
                RUNTIME_CHECK_MSG(original_msg == nullptr || *original_msg->remaining_consumers == 0, "Fine grained receiver pop a message that is not full consumed, remaining consumer: {}", *original_msg->remaining_consumers);
```